src/libsystemd/sd-bus/bus-control.c

```diff
@@ -27,6 +27,7 @@ _public_ int sd_bus_get_unique_name(sd_bus *bus, const char **unique) {
         assert_return(bus = bus_resolve(bus), -ENOPKG);
         assert_return(unique, -EINVAL);
         assert_return(!bus_pid_changed(bus), -ECHILD);
+        assert_return(!bus_nonce_changed(bus), -EUCLEAN);
```

Comment:
hmm, i'd prefer if we could have a single concept for this, that checks both the compilation module and the pid in one.

i.e. just add a helper:

```c
static uint64_t origin_nonce;

static void origin_nonce_initialize(void) {
        origin_nonce = random_u64();
}

static uint64_t origin_id(void) {
        static pthread_once_t once = PTHREAD_ONCE_INIT;
        assert_se(pthread_once(&once, bus_object_nonce_initialize) == 0);
        return origin_nonce ^ getpid_cached();
}
```

And then replace the current "original_pid" field in "sd_bus" with "origin_id" that is just initialized from origin_id().

And simply return ECHILD in case the origin ID changes. Yes, the error code is a bit less accurate than your patch here, but conceptually it's much simpler I think.

Also, the above code should probably be wrapped in some macro, since we probably want this in sd-event too, and elsehwere where we currently have pid_changed() function. sd-journal, …